project/src/com/google/daggerquery/executor/services/SourcesLoader.java

```diff
@@ -16,24 +16,73 @@
 
 package com.google.daggerquery.executor.services;
 
+import com.google.common.io.Files;
 import com.google.daggerquery.protobuf.autogen.BindingGraphProto.BindingGraph;
+import java.io.File;
 import java.io.FileNotFoundException;
 import java.io.IOException;
 import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
 
 /**
  * A class which loads a binding graph saved with Dagger SPI plugin.
  */
 public class SourcesLoader {
-  private static final String PATH_TO_BINDING_GRAPH = "/com/google/daggerquery/binding_graph_data.textproto";
+  private static final String PATH_TO_BINDING_GRAPHS = "/com/google/daggerquery/binding_graph_data.zip";
+  private static final String BINDING_GRAPHS_SOURCES = "binding_graphs.bin";
 
-  public BindingGraph loadBindingGraph() throws IOException {
-    InputStream inputStream = SourcesLoader.class.getResourceAsStream(PATH_TO_BINDING_GRAPH);
+  /**
+   * Reads .zip resource file which contains several .textproto files with serialized binding graphs.
+   *
+   * <p>Returns a list with {@link BindingGraph} instances. Parses them from files located in .zip file.
+   *
+   * @throws FileNotFoundException if an app with the connected plugin wasn't launched and .zip file cannot be found
+   * @throws IOException if an I/O error occurred while extracting files from .zip file
+   */
+  public List<BindingGraph> loadBindingGraphs() throws IOException {
+    try (InputStream zipInputStream = SourcesLoader.class.getResourceAsStream(PATH_TO_BINDING_GRAPHS)) {
+      if (zipInputStream == null) {
+        throw new FileNotFoundException(String.format("File %s is missing.", PATH_TO_BINDING_GRAPHS));
+      }
 
-    if (inputStream == null) {
-      throw new FileNotFoundException(String.format("File %s is missing.", PATH_TO_BINDING_GRAPH));
+      ZipFile zipFile = makeZipFileFromInputStream(zipInputStream);
+      Enumeration<? extends ZipEntry> filesWithBindingGraphs = zipFile.entries();
```

Comment:
Instead of having to define your helper method, would looping directly over [`ZipInputStream.getNextEntry()`](https://docs.oracle.com/javase/8/docs/api/java/util/zip/ZipInputStream.html#getNextEntry--) work?

Reply:
I thought about it but I didn't find another easier way to read the content of .textproto files from .zip. For it now I use the method `getInputStream(...)` from `ZipFile` instance.

Reply:
Sorry maybe I wasn't clear - the `makeZipFileFromInputStream` helper method seems to just convert the `ZipInputStream` into `ZipFile`, and then `zipFile.entries()` is called to obtain `Enumeration<? extends ZipEntry>`. Based on the type signature I would have thought iterating through `ZipInputStream.getNextEntry()` would be equivalent to `Iterable<ZipEntry>`. I haven't tried this myself so I could be wrong...

Reply:
Yes, I got it. I can use it for getting entries this way: 

```
ZipEntry bindingGraphEntry = zipInputStream.getNextEntry();
while (bindingGraphEntry != null) {

    try (InputStream bindingGraphInputStream = zipFile.getInputStream(bindingGraphEntry)) {
        bindingGraphs.add(BindingGraph.parseFrom(bindingGraphInputStream));
    }

    bindingGraphEntry = zipInputStream.getNextEntry();
}
```

The problem is that now I use `ZipFile` instance also in a line 59, that's why it's hard to get rid of this helper method.

Reply:
Oh I see, yeah you taught me something new :smiley: